examples/chef/common/stubs.cpp

```diff
@@ -8,6 +8,161 @@
 #ifdef EMBER_AF_PLUGIN_DOOR_LOCK_SERVER
 #include <app/clusters/door-lock-server/door-lock-server.h>
 
+class LockManager
+{
+public:
+    static constexpr uint32_t kNumEndpoints              = 1;
+    static constexpr uint32_t kNumUsersPerEndpoint       = 2;
+    static constexpr uint32_t kNumCredentialsPerEndpoint = 20;
+    static constexpr uint32_t kNumCredentialsPerUser     = 10;
+    static constexpr uint32_t kMaxNameLength             = 32;
+    static constexpr uint32_t kMaxDataLength             = 16;
+
+    struct Credential
+    {
+        bool set(DlCredentialStatus status, DlCredentialType type, chip::ByteSpan newData)
+        {
+            if (newData.size() > kMaxDataLength)
+                return false;
+            memcpy(data, newData.data(), newData.size());
+            info = EmberAfPluginDoorLockCredentialInfo{
+                status,
+                type,
+                chip::ByteSpan(data, newData.size()),
+            };
+            return true;
+        }
+
+        EmberAfPluginDoorLockCredentialInfo info = { DlCredentialStatus::kAvailable };
+        uint8_t data[kMaxDataLength];
+    };
+
+    struct User
+    {
+        void set(chip::CharSpan newName, uint32_t userId, DlUserStatus userStatus, DlUserType type, DlCredentialRule credentialRule)
+        {
+            size_t sz = std::min(sizeof(name), newName.size());
+            memcpy(name, newName.data(), sz);
+            info = EmberAfPluginDoorLockUserInfo{
+                chip::CharSpan(name, sz), chip::Span<const DlCredential>(), userId, userStatus, type, credentialRule,
+            };
+        }
+        bool addCredential(uint8_t type, uint16_t index)
+        {
+            if (info.credentials.size() == kNumCredentialsPerUser)
+                return false;
+            auto & cr          = credentialMap[info.credentials.size()];
+            cr.CredentialType  = type;
+            cr.CredentialIndex = index;
+            info.credentials   = chip::Span<const DlCredential>(credentialMap, info.credentials.size() + 1);
+            return true;
+        }
+
+        EmberAfPluginDoorLockUserInfo info = { .userStatus = DlUserStatus::kAvailable };
+        char name[kMaxNameLength];
+        DlCredential credentialMap[kNumCredentialsPerUser];
+    };
+
+    struct Endpoint
+    {
+        chip::EndpointId id;
+        User users[kNumUsersPerEndpoint];
+        Credential credentials[(int) DlCredentialType::kUnknownEnumValue][kNumCredentialsPerEndpoint];
```

Comment:
Please don't assume the unknown value will be equal to the number of valid values.  It can be something else entirely, depending on what values are defined.  In particular, nothing requires credential type values to be contiguous, or the unknown value to come after the last defined value.

In practice, you should figure out what credential types you support and what indices they are stored at and size accordingly....

Reply:
Okay well this sample is really only interested in PINs so dropping back down to a single array